### PR TITLE
Add globalization

### PR DESCRIPTION
**Pull request by superkhau**
Connect to https://github.com/strongloop/loopback/issues/2422

### CODE REVIEW

**Comment by superkhau** [comment]
@slnode test please

File: lib/providers/apns.js [hunk 1]
```diff
@@ -83,7 +84,7 @@ ApnsProvider.prototype._setupPushConnection = function(options) {
   connection.on('socketError', errorHandler);
 
   connection.on('transmissionError', function(code, notification, recipient) {
-    var err = new Error('Cannot send APNS notification: ' + code);
+    var err = new Error(g.f('Cannot send APNS notification: %s', code));
```
Comment (0candy) [inline lib/providers/apns.js]: We don't need translation for `APNS` put `{{}}` around it.
`{{APNS}}`

File: lib/providers/gcm.js [hunk 2]
```diff
@@ -39,7 +41,8 @@ GcmProvider.prototype.pushNotification = function(notification, deviceToken) {
           debug('Device %j is no longer registered.', registrationIds[index]);
           devicesGoneRegistrationIds.push(registrationIds[index]);
         } else if (code) {
-          errors.push('GCM error code: ' + (code || 'Unknown') + ', deviceToken: ' + registrationIds[index]);
+          errors.push(g.f('GCM error code: %s, deviceToken: %s',
```
Comment (0candy) [inline lib/providers/gcm.js]: `{{GCM}}`

File: lib/payload.js [hunk 3]
```diff
@@ -26,11 +28,12 @@ function Payload(data) {
         if (!__hasProp.call(data, key)) continue;
         value = data[key];
         if (typeof key !== 'string' || key.length === 0) {
-            throw new Error("Invalid field (empty)");
+            throw new Error(g.t("Invalid field (empty)"));
```
Comment (0candy) [inline lib/payload.js]: Can you change it to `'` single quotes?

File: lib/payload.js [hunk 4]
```diff
@@ -26,11 +28,12 @@ function Payload(data) {
...
         }
         if (typeof value !== 'string') {
-            throw new Error("Invalid value for `" + key + "'");
+            throw new Error(g.f("Invalid value for `%s`", key));
```
Comment (0candy) [inline lib/payload.js]: same here

File: lib/payload.js [hunk 5]
```diff
@@ -45,7 +48,7 @@ function Payload(data) {
                 if ((_ref = key.split('.', 2), prefix = _ref[0], subkey = _ref[1], _ref).length === 2) {
                     this[prefix][subkey] = value;
                 } else {
-                    throw new Error("Invalid field: " + key);
+                    throw new Error(g.f("Invalid field: %s", key));
```
Comment (0candy) [inline lib/payload.js]: same

File: lib/payload.js [hunk 6]
```diff
@@ -118,15 +121,15 @@ Payload.prototype.variable = function (keyPath) {
         if ((_ref = this.event) != null ? _ref.name : undefined) {
             return (_ref1 = this.event) != null ? _ref1.name : undefined;
         } else {
-            throw new Error("The ${" + keyPath + "} does not exist");
+            throw new Error(g.f("The ${%s} does not exist", keyPath));
```
Comment (0candy) [inline lib/payload.js]: same

File: lib/payload.js [hunk 7]
```diff
@@ -118,15 +121,15 @@ Payload.prototype.variable = function (keyPath) {
...
         }
     }
     _ref2 = keyPath.split('.', 2), prefix = _ref2[0], key = _ref2[1];
     if (prefix !== 'var' && prefix !== 'data') {
-        throw new Error("Invalid variable type for ${" + keyPath + "}");
+        throw new Error(g.f("Invalid variable type for ${%s}", keyPath));
```
Comment (0candy) [inline lib/payload.js]: same

File: lib/payload.js [hunk 8]
```diff
@@ -118,15 +121,15 @@ Payload.prototype.variable = function (keyPath) {
...
     }
     if (this[prefix][key] == null) {
-        throw new Error("The ${" + keyPath + "} does not exist");
+        throw new Error(g.f("The ${%s} does not exist", keyPath));
```
Comment (0candy) [inline lib/payload.js]: same

File: lib/payload.js (hunk 3 above)
Comment (superkhau) [inline lib/payload.js]: That would make it an unrelated change. That should be done in a separate PR.

**Comment by superkhau** [comment]
@0candy Review again please.

File: lib/providers/apns.js [hunk 9]
```diff
@@ -83,7 +84,7 @@ ApnsProvider.prototype._setupPushConnection = function(options) {
   connection.on('socketError', errorHandler);
 
   connection.on('transmissionError', function(code, notification, recipient) {
-    var err = new Error('Cannot send APNS notification: ' + code);
+    var err = new Error(g.f('Cannot send {{APNS}} notification: %s', code));
```
Comment (0candy) [inline lib/providers/apns.js]: I think you have to run `slt-globalize -e` again for `intl/en/messages.json` to get updated again with the brackets change.

**Comment by superkhau** [comment]
@0candy Updated, PTAL again.

**Comment by 0candy** [comment]
The change itself LGTM.
But there is a globalize error right now, all the CI test fail.
```assert.js:86
  throw new assert.AssertionError({
        ^
AssertionError: Not supported: en
    at loadGlobalize (/mnt/jenkins/workspace/loopback-component-push/cdcdef7b/node_modules/strong-globalize/lib/globalize.js:354:3)
    at Object.setDefaultLanguage (/mnt/jenkins/workspace/loopback-component-push/cdcdef7b/node_modules/strong-globalize/lib/globalize.js:100:3)
    at new StrongGlobalize (/mnt/jenkins/workspace/loopback-component-push/cdcdef7b/node_modules/strong-globalize/index.js:27:41)
    at StrongGlobalize (/mnt/jenkins/workspace/loopback-component-push/cdcdef7b/node_modules/strong-globalize/index.js:25:12)
    at Object.<anonymous> (/mnt/jenkins/workspace/loopback-component-push/cdcdef7b/lib/providers/apns.js:11:9)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
```
So please don't merge yet.

File: index.js [hunk 10]
```diff
@@ -9,6 +9,8 @@
 var loopback = require('loopback');
 var PushConnector = require('./lib/push-connector');
 exports = module.exports = PushConnector;
+var SG = require('strong-globalize');
```
Comment (Setogit) [inline index.js]: Why not following strong-globalize coding guideline here?
https://www.npmjs.com/package/strong-globalize#autonomous-message-loading
Please study paragraphs after "For example, the following does not work as intended because the package sub calls SG.SetRootDir first:"

File: lib/payload.js [hunk 11]
```diff
@@ -6,6 +6,8 @@
 // this file has no tests so avoid refactor and ignore jshint for now
 /*jshint ignore: start */
 var serial,__hasProp = {}.hasOwnProperty;
+var SG = require('strong-globalize');
```
Comment (Setogit) [inline lib/payload.js]: Why not following strong-globalize coding guideline here?
https://www.npmjs.com/package/strong-globalize#autonomous-message-loading
Please study paragraphs after "For example, the following does not work as intended because the package sub calls SG.SetRootDir first:"

File: lib/providers/gcm.js [hunk 12]
```diff
@@ -8,6 +8,8 @@ var extend = require('util')._extend;
 var EventEmitter = require('events').EventEmitter;
 var gcm = require('node-gcm');
 var debug = require('debug')('loopback:component:push:provider:gcm');
+var SG = require('strong-globalize');
```
Comment (Setogit) [inline lib/providers/gcm.js]: Why not following strong-globalize coding guideline here?
https://www.npmjs.com/package/strong-globalize#autonomous-message-loading
Please study paragraphs after "For example, the following does not work as intended because the package sub calls SG.SetRootDir first:"

File: lib/push-manager.js [hunk 13]
```diff
@@ -12,6 +12,8 @@ var providers = require('./providers');
 var loopback = require('loopback');
 var NodeCache = require('node-cache');
 var debug = require('debug')('loopback:component:push:push-manager');
+var SG = require('strong-globalize');
```
Comment (Setogit) [inline lib/push-manager.js]: Why not following strong-globalize coding guideline here?
https://www.npmjs.com/package/strong-globalize#autonomous-message-loading
Please study paragraphs after "For example, the following does not work as intended because the package sub calls SG.SetRootDir first:"

**Comment by Setogit** [comment]
For each JS module, please examine where `strong-globalize' is required and make sure the strong-globalize coding guideline is followed.
https://www.npmjs.com/package/strong-globalize#autonomous-message-loading
Paragraphs after "For example, the following does not work as intended because the package sub calls SG.SetRootDir first:"

**Comment by Setogit** [comment]
test please

**Comment by 0candy** [comment]
ah sorry, missed that. Please fix Tetsuo's comment.

**Comment by rmg** [comment]
@slnode test please
